### PR TITLE
feat : add fenwick tree 2D data structure, struct fenwick2D

### PR DESCRIPTION
**Pull request by zyrch**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1230"></a>

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1230"></a>

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1230"></a>

### CODE REVIEW

**Comment by kvedala** [comment]
Review existing code and standards: https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/range_queries/fenwick_tree.cpp
image

**Comment by zyrch** [comment]
Changes have been made.

**Review by ayaankhan98 — APPROVED** [review]

**Comment by ayaankhan98** [comment]
code looks fine but documentation is not up to standard

**Comment by zyrch** [comment]
changes have been made

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Documentation is still not up to the [repository standards](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md).
Please see the [typical structure](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) of a program, that might help. 🙂

**Comment by realstealthninja** [comment]
this seems to be stale

File: range_queries/fenwick_tree_2d.cpp [hunk 1]
```diff
@@ -0,0 +1,129 @@
+#include <iostream>
+#include <cassert>
+#include <vector>
```
Comment (realstealthninja) [inline range_queries/fenwick_tree_2d.cpp:3]:
```suggestion
#include <iostream> /// for I/O operations
#include <cassert> /// for assert
#include <vector> /// for std::vector
```
needs documentation

File: range_queries/fenwick_tree_2d.cpp [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    fenwick_tree_2D.update(3, 2, 5);
+    fenwick_tree_2D.update(3, 3, 6);
+
+    assert(fenwick_tree_2D.query(1, 1, 2, 2) == 12);
+    assert(fenwick_tree_2D.query(1, 2, 3, 3) == 27);
+    fenwick_tree_2D.update(1, 1, -3);
+    assert(fenwick_tree_2D.query(1, 1, 2, 2) == 9);
+    assert(fenwick_tree_2D.query(1, 2, 3, 3) == 27);
+
+    std::cout << "All Test Passed\n";
```
Comment (realstealthninja) [inline range_queries/fenwick_tree_2d.cpp:126]: should be moved to a function
```cpp
void test() {};
```

File: range_queries/fenwick_tree_2d.cpp [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    // sum of region {x1 to x} and {y1 to y}
+    // sum in region x1 to x and y1 to y can written as
+    //    - sum (x, y) - sum(x1 - 1, y) - sum(x, y - 1) + sum(x1 - 1, y1 - 1)
+    // last part as is add rectangle {x1, y1} is subtracted twice
+    T query(int x1, int y1, int x, int y) {
+	return sum(x, y) - sum(x1 - 1, y) - sum(x, y1 - 1) + sum(x1 - 1, y1 - 1);
+    } 
+
+};
+
```
Comment (realstealthninja) [inline range_queries/fenwick_tree_2d.cpp:102]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
```

File: range_queries/fenwick_tree_2d.cpp [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    // n - size of the array on which fenwick tree is used
+    // creates a new array with all entries as 0
+    // indicies are 1-indexed
+    explicit fenwick(int n) {
+	siz = n;
+	arr.resize(n + 1);
+    }
+
+    // Calculates a prefix sum till index idx
+    T sum(int idx) {
```
Comment (fhlasek) [inline range_queries/fenwick_tree_2d.cpp:23]: might want to check if `idx <= siz`

File: range_queries/fenwick_tree_2d.cpp [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	    an += arr[idx];
+	    // subtract first active bit from idx
+	    idx -= idx & -idx;
+	}
+	return an;
+    }
+
+    // update the values at idx by delta
+    // update all the indices which include idx in their sum
+    void update(int idx, T delta) {
```
Comment (fhlasek) [inline range_queries/fenwick_tree_2d.cpp:35]: similarly, could check the bounds of `idx`

File: range_queries/fenwick_tree_2d.cpp [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	while (idx < siz + 1) {
+	    arr[idx] += delta;
+	    // add the first active bit to idx
+	    idx += idx & -idx;
+	}
+    }
+
+    // find sum in range l to r
+    T query(int l, int r) {
```
Comment (fhlasek) [inline range_queries/fenwick_tree_2d.cpp:44]: finally, here are even more implicit assumptions on the input, `0 < l <= r <= siz` (I believe)

File: range_queries/fenwick_tree_2d.cpp [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+	while(x > 0) {
+	    // for each fenwick tree in range calucate sum till y
+	    an += arr[x].sum(y);
+	    // remove the last bit of x from x
+	    x -= x & -x;
+	}
+	return an;
+    }
+
+    // sum of region {x1 to x} and {y1 to y}
```
Comment (fhlasek) [inline range_queries/fenwick_tree_2d.cpp:93]: nit: For clarity, I think it would be better to call them `x1`, `x2` and `y1`, `y2`

**Pull request closed without merging**